### DISCUSSION TITLE
Potential Nix Flake dev environment

### DISCUSSION THREAD

**Discussion by oldaccountdeadname** [qutebrowser/qutebrowser#6804] [Ideas]
Hi! I'm beginning to look at the source code to hack some stuff into my build of qute --- setting up the development environment looks to be as it usually is with Python; that is, more than a bit convoluted for people like me with only passing familiarity with the common build systems. If I were to write a Nix Flake for this, providing a stable development shell and maybe building/packaging, would y'all maintainers be interested in merging it upstream? Ideally, this provides all the standard benefits of Nix: isolation, reproducibility, and stability, with all the standard costs: time and weight[^1].

This is something I'm going to do (and potentially maintain) anyway, so I just wanted to see if it's something that I should a) put actual effort into doing well, and b) send a PR for. I definitely don't want to add undue load to the PRs for something that nobody wants, but I'd be happy to contribute upstream if it's helpful - thanks! 

[^1]: Nix Flakes are also technically experimental, but the interfaces don't look like they'll change that much: https://discourse.nixos.org/t/nix-2-4-released/15822

**Comment by The-Compiler** [comment]
I think I'd prefer it to live in a separate repository, like e.g. the [Debian package](https://github.com/qutebrowser/qutebrowser-debian) or [Flatpak package](https://github.com/flathub/org.qutebrowser.qutebrowser) (and many other distro packages) already do.  Packaging infrastructure other than the official releases (venv stuff and Windows/macOS binaries) usually is in a separate place already for existing packages.

What would be the benefits of having it in the main tree instead?

**Comment by oldaccountdeadname** [comment]
> ... Packaging infrastructure other than the official releases (venv stuff and
> Windows/macOS binaries) usually is in a separate place already for existing
> packages.

I'm honestly not sure whether this is best categorized as packaging
infrastructure. Note that a derivation for (Nix package of) Qutebrowser already
exists outside of this repo[0], and I doubt it would be productive to replicate
it here. Rather, I'm suggesting something more along the lines of general
development infrastructure, very similar to the existing mkvenv.py script, but
encapsulating not just the Python environment but *also* system libraries, etc.

> What would be the benefits of having it in the main tree instead?

Well, the standard Nix workflow is to clone a project, run `nix develop`, and
automatically pull in the necessary dependencies (whether packages, enviroment
variables, etc) and get a fully capable dev environment in a nested shell. This
workflow just takes a bit longer if you need to clone another repo, run `nix
develop`, and then once more cd to qutebrowser's sources. Some Nix users also
like to automatically run `nix-shell` (or, in this case, the equivalent-ish nix
develop) whenever they cd to a directory that allows it, which, obviously,
doesn't transfer over as seamlessly if we're talking about cloning a *different*
repo for this.

In addition, we might want the Flake to read the dependencies from the
requirements.txt's in this repo, which, again, is significantly more complex if
we're not in this repo already.

Despite the paragraphs above, though, I'm honestly fine either way. I'll do
whatever you think would keep everything most organized!

[0]: https://github.com/NixOS/nixpkgs/blob/nixos-21.05/pkgs/applications/networking/browsers/qutebrowser/default.nix

--
lincoln auster
they/them
